apps/server/src/modules/files-storage/service/preview.service.ts

```diff
@@ -0,0 +1,154 @@
+import { Injectable, NotAcceptableException, NotFoundException, UnprocessableEntityException } from '@nestjs/common';
+import { LegacyLogger } from '@src/core/logger';
+import crypto from 'crypto';
+import { subClass } from 'gm';
+import { PassThrough } from 'stream';
+import { S3ClientAdapter } from '../client/s3-client.adapter';
+import { DownloadFileParams, PreviewParams } from '../controller/dto';
+import { FileRecord } from '../entity';
+import { ErrorType } from '../error';
+import { IGetFile, IGetFileResponse } from '../interface';
+import { PreviewOutputMimeTypes } from '../interface/preview-output-mime-types.enum';
+import { FileDtoBuilder, FileResponseBuilder } from '../mapper';
+import { FilesStorageService } from './files-storage.service';
+
+interface PreviewFileParams {
+	fileRecord: FileRecord;
+	downloadParams: DownloadFileParams;
+	previewParams: PreviewParams;
+	hash: string;
+	filePath: string;
+	bytesRange?: string;
+}
+
+@Injectable()
+export class PreviewService {
+	constructor(
+		private readonly storageClient: S3ClientAdapter,
+		private readonly fileStorageService: FilesStorageService,
+		private logger: LegacyLogger
+	) {
+		this.logger.setContext(PreviewService.name);
+	}
+
+	public async getPreview(
+		fileRecord: FileRecord,
+		downloadParams: DownloadFileParams,
+		previewParams: PreviewParams,
+		bytesRange?: string
+	): Promise<IGetFileResponse> {
+		this.checkIfPreviewPossible(fileRecord);
+
+		const { forceUpdate, outputFormat } = previewParams;
+		const hash = this.createNameHash(downloadParams, previewParams);
+		const filePath = this.getFilePath(fileRecord, hash);
+		const name = this.getPreviewName(fileRecord, outputFormat);
+		let file: IGetFile;
+
+		const previewFileParams = { fileRecord, downloadParams, previewParams, hash, filePath, bytesRange };
+
+		if (forceUpdate) {
+			file = await this.generatePreview(previewFileParams);
+		} else {
+			file = await this.tryGetPreviewOrGenerate(previewFileParams);
+		}
+
+		const response = FileResponseBuilder.build(file, name);
+
+		return response;
+	}
+
+	private checkIfPreviewPossible(fileRecord: FileRecord): void | NotAcceptableException {
```

Comment:
This is a good example why I am against using exceptions as return types: You forgot to update it and the IDE does not recognize it. But it is not even the strongest reason for me to be against it. I wanted to write something in the team chat when the discussion last came up and forgot to. I'm gonna do that now.